### ISSUE TITLE
Pass custom environmental variables into Kong ingress controller

### ISSUE THREAD

**Issue by MatteoInfi** [Kong/charts#411]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem Statement

So, I need to pass extra env variable to my custom plugin.
In docker compose I just do this:

```
- DATABASE_HOST=myhost
- DATABASE_PASSWORD=123
- DATABASE_USER=root
- DATABASE_NAME=kong
- "KONG_NGINX_MAIN_ENV=DATABASE_HOST; env DATABASE_PASSWORD; env DATABASE_USER; env DATABASE_NAME"
```

This is working fine, however I don't know how to do the same with the Kong ingress controller, according to the chart value description:

> Specify Kong configuration
This chart takes all entries defined under `.env` and transforms them into into `KONG_*`
environment variables for Kong containers.
Their names here should match the names used in https://github.com/Kong/kong/blob/master/kong.conf.default
See https://docs.konghq.com/latest/configuration also for additional details

```
  env:
    log_level: info
    my_custom_var: hello
```

So if I put my custom variable there it will be transformed into `KONG_MY_CUSTOM_VAR`

So how can I do it? I cannot find where the Kong chart exposes additional environmental variables for the proxy pod

## Possible solutions
Just rename the variables this way:

```
- KONG_DATABASE_HOST=myhost
- KONG_DATABASE_PASSWORD=123
- "KONG_NGINX_MAIN_ENV=KONG_DATABASE_HOST; env KONG_DATABASE_PASSWORD"
```

---

Or override Kong deployment:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ingress-kong
  namespace: kong
spec:
  template:
    spec:
      containers:
      - name: proxy
        env:
        - name: DATABASE_HOST
          value: myhost
```

Is there a better way to do this?

Thanks

**Comment by shaneutt** [comment]
Using the `.env` settings will as you suggest result in the env var being prefixed with `KONG_`, this was a deliberate choice in the helper script:

https://github.com/Kong/charts/blob/main/charts/kong/templates/_helpers.tpl#L752

```
{{/*
    ====== USER-SET ENVIRONMENT VARIABLES ======
*/}}
{{- $userEnv := dict -}}
{{- range $key, $val := .Values.env }}
  {{- $upper := upper $key -}}
  {{- $var := printf "KONG_%s" $upper -}}
  {{- $_ := set $userEnv $var $val -}}
{{- end -}}
```

It would be possible however to add additional functionality to support your use case, let me know if you want to follow up on this as a feature request?

As for a workaround for this issue in the meantime [unfortunately helm itself doesn't have a means to inject env vars into containers](https://github.com/helm/helm/issues/7610) so your options are limited to post-processing workarounds (e.g. manually adding the ENV vars you want to your pods, or adding a initContainer to do it) or to clone this repo and locally modify the chart and deploy using your own modified version until such a time that there's a supported feature.

Let me know your thoughts, if that is helpful, and any other questions you may have?

**Comment by MatteoInfi** [comment]
So, is it not possible to pass env variables to a custom plugin via Helm chart? It seems a common use case, or maybe is there something I am missing?

In this case we could do something similar to the Bitnami chart: https://github.com/bitnami/charts/tree/master/bitnami/postgresql#postgresql-parameters
have the possibility to pass extra env variables to a pod via: `extraEnv` 

```
extraEnv:
 MY_CUSTOM_VARIABLE: hello // either as map or array of objects
```

And then merge it together with the other variables: 

```
{{- $extraEnv := dict -}}
{{- range $key, $val := .Values.extraEnv }} // Maybe a check to see whether the value is actually being set
  ...
{{- end -}}

{{/*
      ====== MERGE AND RENDER ENV BLOCK ======
*/}}

{{- $completeEnv := mergeOverwrite $customEnv $userEnv -}}
{{- template "kong.renderEnv" $completeEnv -}}

{{- end -}}
```

Not sure how `mergeOverwrite` works, but that would be the idea

What do you think?

**Comment by shaneutt** [comment]
Yeah unfortunately so far we just didn't have this.

Your suggestion looks good I talked it over with my fellow maintainers and we would accept that kind of patch. :+1:

**Comment by MatteoGioioso** [comment]
Hello,
I have the same requirement. 
@shaneutt if you guys are accepting a PR, I will be glad to try and work something out as we need this feature. 

Please let me know, thanks.

**Comment by shaneutt** [comment]
> Hello,
> I have the same requirement.
> @shaneutt if you guys are accepting a PR, I will be glad to try and work something out as we need this feature.
> 
> Please let me know, thanks.

Yep sounds good thanks, I would recommend coordinating with @MatteoInfi prior to starting but then go for it.

**Comment by MatteoInfi** [comment]
I honestly do not have much time and I am not really expert with helm templating, if you can do it is better.
Thanks

**Comment by MatteoGioioso** [comment]
Gotcha, I will see what I can do next week

**Comment by JuanMorenoDeveloper** [comment]
Related to this https://github.com/Kong/charts/pull/568, This is available since v 2.8.0
